include/swift/SIL/SILFunction.h

```diff
@@ -573,13 +573,27 @@ class SILFunction
 
   /// Get this function's thunk attribute.
   IsThunk_t isThunk() const { return IsThunk_t(Thunk); }
-  void setThunk(IsThunk_t isThunk) { Thunk = isThunk; }
+  void setThunk(IsThunk_t isThunk, bool forSignatureOptimization = false) {
+    // The _original_ function for a method can turn into a thunk through
+    // signature optimization, meaning it needs to retain its subclassScope
+    assert(getClassSubclassScope() == SubclassScope::NotApplicable ||
```

Comment:
This seems overkill. If you're intent on keeping this, you should add a new case to IsThunk_t for signature-optimized thunks. A future assertion could also ensure that any thunk other than a signature thunk is shared or private, etc.